src/git/loose.ml

```diff
@@ -199,47 +199,29 @@ module Make
           | Error _ -> acc)
         [] firsts
 
-  type 't decoder =
-    (module S.DECODER
-      with type t = 't
-       and type init = Inflate.window * Cstruct.t * Cstruct.t
-       and type error = [ Error.Decoder.t
-                        | `Inflate of Inflate.error ])
-
-  let gen (type t) ~fs ~root ~window ~ztmp ~dtmp ~raw (decoder : t decoder) hash =
+  type ('result, 'decoder) decoder =
+    (module Helper.DECODER
+            with type t = 'result
+             and type decoder = 'decoder
+             and type error = [ Error.Decoder.t | `Inflate of Inflate.error ] )
+
+  let gen (type state) (type result) ~fs ~root (state:state) ~raw (decoder : (result, state) decoder) hash =
     let module D = (val decoder) in
+    let module Decoder = Helper.Decoder(D)(FS) in
     let first, rest = explode hash in
-    let decoder     = D.default (window, ztmp, dtmp) in
-    let path        = Fpath.(root / "objects" / first / rest) in
-    Log.debug (fun l -> l "Reading the loose object %a." Fpath.pp path);
-    FS.with_open_r fs path @@ fun ic ->
-    let rec loop decoder =
-      match D.eval decoder with
-      | `Error (_, (#Error.Decoder.t as err)) ->
-        Lwt.return Error.(v @@ Error.Decoder.with_path path err)
-      | `Error (_, (`Inflate _ as err)) ->
-        Lwt.return Error.(v @@ Inf.with_path path err)
-      | `End (_, value) -> Lwt.return (Ok value)
-      | `Await decoder ->
-        FS.File.read raw ic >>= function
-        | Error err -> Lwt.return Error.(v @@ FS.err_read path err)
-        | Ok n ->
-          Log.debug (fun l ->
-              l "Reading %d byte(s) of the file-descriptor (object: %a)."
-                n Hash.pp hash);
-          match D.refill (Cstruct.sub raw 0 n) decoder with
-          | Ok decoder -> loop decoder
-          | Error (#Error.Decoder.t as err) ->
-            Lwt.return Error.(v @@ Decoder.with_path path err)
-          | Error (`Inflate _ as err) ->
-            Lwt.return Error.(v @@ Inf.with_path path err)
-    in
-    loop decoder >|= fun ret ->
-    Log.debug (fun l -> l "Finished to read the object %s / %s." first rest);
-    ret
+    let file        = Fpath.(root / "objects" / first / rest) in
+    Log.debug (fun l -> l "Reading the loose object %a." Fpath.pp file);
+    Decoder.of_file fs file raw state >>= function
+    | Ok v -> Lwt.return (Ok v)
+    | Error (`Decoder (#Error.Decoder.t as err)) ->
```

Comment:
do you really need to specify the error type here? Would something like work?

```
| Error (`Decoder err) -> Error (v @@ Error.Decoder.with_path file err)
```

Reply:
Only on this case (about `packed-refs` and `reference`, we don't have this problem), `Decoder` can have 2 kinds of error:
- `Inflate`: about `inflate` implementation
- `Error.Decoder.t`: usual decoder error

OCaml seems to be not so clever about pattern-matching and polymorphic variant. So, if I don't use `#Error.Decoder.t`, it includes `Inflate` error. However, `Error.Decoder.with_path` does not expect this kind of error. So, yeah :( I need to refine pattern-matching.

About `packed-refs` and `reference`, because we don't have `Inflate` error, I can change it, of course (more readable).